pub/README.md

```diff
@@ -4,14 +4,16 @@ Dart (pub) support for [`dependabot-core`][core-repo].
 
 ### Limitations
 
- - No support for updating git-dependencies
-   * `dart pub` in general doesn't read versions from git, so upgrade logic is limited.
-   * Some variant of support for updating git-dependencies could be added in the future.
+ - Limited updating of git-dependencies
+   * `dart pub` in general doesn't read versions numbers from git, so upgrade logic is limited to upgrading to what the 'ref' is pointing to.
+   * If you pin to a specific revision in `pubspec.yaml` dependabot will not find upgrades.
+   * If you give a branch in `pubspec.yaml` dependabot will upgrade to the
+     latest revision that branch is pointing to, and update pubspec.lock
+     accordingly.
  - No support for security advisory integration.
  - If the version found is ignored (by dependabot config) no update will happen (even if, an earlier version could be used)
- - Very limited metadata support (just retrieves the github link)
+ - Limited metadata support (just retrieves the repository link).
  - Only stable versions of Dart and Flutter supported.
- - No support for private package repositories (mostly a configuration issue).
```

Comment:
I believe we do support this by using `SharedHelpers.with_git_configured` here: [SharedHelpers.with_git_configured(credentials: credentials) ](https://github.com/dependabot/dependabot-core/blob/main/pub/lib/dependabot/pub/helpers.rb#L52)

I did a local test using the dry-run script of one private repo having a git dependency on another private repo, and it seemed to work.

I might be misunderstanding something - is there more we need to do to support this?

Reply:
Git dependencies should work but I was wondering if this was referring to running a private package repo as described here: https://dart.dev/tools/pub/custom-package-repositories

It looks like this adds support for calling out to the rest api of additional repos besides `https://pub.dev` but I was wondering if we'd need to authenticate with them?

Reply:
On the dependabot-core side we provide a list of credentials to the UpdateChecker/FileUpdater ([example](https://github.com/dependabot/dependabot-core/blob/c6f81a6cab8e075893b40cd740b62fba2636d52d/common/lib/dependabot/update_checkers/base.rb#L23)). We'd have to agree on what the format would be but it would be similar to:

```
[
  {
    "type": "pub_repository",
    "url": "https://pub.me",
    "token": "...."
   }
]
```

You could then configure pub with those credentials if present.

However, there's still some work we'd need to do on our side to allow those to be configured in the `dependabot.yml` and the way credentials are applied when running on GitHub is a bit different so we'll need adjustments there as well. This wouldn't be a high priority for us to implement right now although that could change if we see a lot of interest in supporting it.

Reply:
I think adding support for custom repositories is simple, if we have:
```js
[
  {
    "type": "pub_repository",
    "url": "https://pub.me",
    "token": "some-super-secret-string"
   }
]
```
Then before running dependency_services, I think we can simply do:
```bash
dart pub token add https://pub.me --env-var PUB_TOKEN_1
```

Then when running dependency services we run with environment variables like:
 * `PUB_TOKEN_1=some-super-secret-string`

This ensures that secrets aren't cached on disk.
Though it's probably still preferable to do:
```bash
dart pub token add https://pub.me
```

To ensure that the configured environment variables are removed from the configuration.


-----------------

It's possible that would still require a few tweaks in dependency-services, but those should be fast, I don't know if filtered out custom repositories, or if we just let them fail if authentication is required.

Reply:
@mctofu it might be a good idea to file an issue for custom repository authentication. If you get the configuration up and running on your side, I think we can quickly inject env vars and tweak dependency-services if needed.